octodns_cloudflare/__init__.py

```diff
@@ -1209,23 +1227,77 @@ def _apply_Delete(self, change):
+        raise SupportsException(msg)
+
+    def _update_plan(self, zone_name, plan):
+        current_plan = self.zones[zone_name].get('plan', False)
+        if current_plan and current_plan == plan:
+            return
+        if plan in self._supported_plans(zone_name):
+            zone_id = self.zones[zone_name]['id']
+            data = {'plan': {'legacy_id': plan}}
+            resp = self._try_request('PATCH', f'/zones/{zone_id}', data=data)
+            # Update the cached plan information
+            self.zones[zone_name]['plan'] = resp['result']['plan']['legacy_id']
+        else:
+            msg = f'{self.id}: {plan} is not supported for {zone_name}'
+            raise SupportsException(msg)
+
+    def _plan_meta(self, existing, desired, changes):
+        meta = {}
+        zone_name = desired.name
+        desired_plan = self.plan_type
+        if zone_name in self.zones:
+            current_plan = self.zones[zone_name].get('plan', False)
+            if desired_plan is None and current_plan is not False:
+                # If desired plan is removed and current plan is set, set to free plan
+                meta['cloudflare_plan'] = CLOUDFLARE_FREE_PLAN
+            elif current_plan != desired_plan:
+                meta['cloudflare_plan'] = desired_plan
+        return meta
+
     def _apply(self, plan):
         desired = plan.desired
         changes = plan.changes
+        zone_name = desired.name
+
         self.log.debug(
-            '_apply: zone=%s, len(changes)=%d', desired.name, len(changes)
+            '_apply: zone=%s, len(changes)=%d', zone_name, len(changes)
         )
 
-        name = desired.name
-        if name not in self.zones:
+        if zone_name not in self.zones:
             self.log.debug('_apply:   no matching zone, creating')
-            data = {'name': name[:-1], 'jump_start': False}
+            data = {'name': zone_name[:-1], 'jump_start': False}
             if self.account_id is not None:
                 data['account'] = {'id': self.account_id}
+            if self.plan_type is not None:
+                data['plan'] = {'legacy_id': self.plan_type}
             resp = self._try_request('POST', '/zones', data=data)
             zone_id = resp['result']['id']
-            self.zones[name] = zone_id
-            self._zone_records[name] = {}
+            self.zones[zone_name] = {'id': zone_id}
+            plan = resp['result'].get('plan', {}).get('legacy_id', False)
+            if plan is not False:
+                self.zones[zone_name]['plan'] = plan
+
+        # Handle plan changes if needed
+        # older versions of octodns don't have meta support.
+        if not hasattr(plan, 'meta') and self.plan_type is not None:
+            self.log.warning(
+                'plan_type is set but meta is not supported by octodns %s, plan changes will not be applied',
+                octodns_version,
+            )
+        else:
+            meta = getattr(plan, 'meta', {})
```

Comment:
Since the attribute existence is checked above it can just be used here.

```suggestion
            meta = plan.meta
```